RTD: add a "how to optimize a function" example (#1293)
sherpa/sherpa: docs/extra/optimize_function.rst

********************

> Optimizing functions

Reviewer: I suggest "finding minima" or something like that. When I read it first, I thought this is about optimizing the code to make functions run faster. Maybe I'm too much of a coder an normal users won't think that way.

Reviewer: Maybe "minimizing functions"?

Author: I note that the scipy docs are "Optimization"

https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html

and we have a big section on "Optimisers" at https://sherpa.readthedocs.io/en/latest/optimisers/index.html so I'm less concerned about this issue than you are. I don't particularly like the title but I do not have the energy/time to come up with a better one.

I will add a link to the scipy docs to let users know what we're trying to do here.

Reviewer: OK, good point. Fine with me.

where `cb` is a callable that given a set or parameters returns the
current statistic, `start` is the starting positions of the
parameters, `lolims` and `hilims` are the bounds of the parameters,
and the remaining keyword arguments are specific to the optimizer.

Examples
========

The following imports have been made::

  >>> import numpy as np
  >>> from matplotlib import pyplot as plt
  >>> from matplotlib import colors
  >>> from mpl_toolkits.mplot3d import Axes3D
  >>> from astropy.table import Table
  >>> from sherpa.optmethods import optfcts

A simple function
-----------------

The function we want to optimize is a version of the
`Rosenbrock function <https://en.wikipedia.org/wiki/Rosenbrock_function>`_,
where `a=2` and `b=10`, so the best-fit location is at::

  x = 2
  y = 2 * 2

.. note::
   The Sherpa test suite contains a number of test cases based on
   functions like this, taken from
   `More, J. J., Garbow, B. S., and Hillstrom, K. E.. Testing unconstrained optimization software. United States: N. p., 1978. Web. doi:10.2172/6650344. <https://www.osti.gov/biblio/6650344>`_.

The function is::

  def rosenbrock(x, y):
      a, b = 2, 10

>       return (a - x) * (a - x) + b * (y - x * x) * (y - x * x)

Reviewer: ```suggestion
      return (a - x)**2 + b * (y - x * x)**2
```
personal preference. I think it's easier to read because it's shorter, but either way is fine.

Author: How about

```suggestion
      return (a - x)**2  + b * (y - x**2)**2   # :-)
```

  >>> print(surface.shape)
  (71, 61)
  >>> print(surface.min())
  0.0
  >>> print(surface.max())
  1235.0

This can be visualized as::

  >>> fig, ax = plt.subplots(subplot_kw={"projection": "3d"})
  >>> surf = ax.plot_surface(x, y, surface,
  ...                        cmap=cm.plasma_r, alpha=0.8,
  ...                        vlim=0, vmax=300)
  >>> ax.set_xlabel('x')
  >>> ax.set_ylabel('y')
  >>> ax.view_init(30, -100)

.. image:: ../_static/extra/rosenbrock-surface.png

In order to optimize this function we need to determine how we
want to define the "search surface"; that is the value that the
optimizer is going to try and minimize. For this dataset, where
the minimum is 0, we can use the square of the function (i.e. a
least-squares model). Now, the function we pass to the minimizer
requires two values - the "statistic" and a per-parameter breakdown
of the parameter - but for now we are going to ignore the latter,
as it's only needed for some optimizers. This gives us::

  def to_optimize1(args):
      x = args[0]
      y = args[1]
      pa = (2 - x) * (2 - x)
      pb = 10 * (y - x * x) * (y - x * x)
      stat = pa * pa + pb * pb
      return stat, None

> The function can tehn be monimized by passing the routine, the

Reviewer: ```suggestion
The function can then be minimized by passing the routine, the
```
I know you put that in jsut ot maek suer we pya attentions.

Author: Look. You thought we were learning about how to speed up code, not monomize functions!

Reviewer: monomized, sure. But monimized? https://in.pinterest.com/srsirisharao/monomize/

> starting point, and the bounds, to the optimizer- in this case the

Reviewer: ```suggestion
starting point, and the bounds, to the optimizer - in this case the
```

> :py:func:`~sherpa.optmethods.optfcts.minim` routine:

Reviewer: ```suggestion
`~sherpa.optmethods.optfcts.minim` routine:
```

Should work here, too. But not 100% certain, need to check output at RTD if we go with this suggestion. (If it does not work, I need to tweak the conf setting more - not not for this release.)

Author: Interestingly they are **slightly** different, but not enough that I care about...

```
TEST1 `~sherpa.optmethods.optfcts.minim` routine

TEST2 :py:func:`~sherpa.optmethods.optfcts.minim` routine
```

creates

![image](https://user-images.githubusercontent.com/224638/133847219-fefa185c-7f6f-450d-bb24-4356fc570944.png)

  >>> for method in [optfcts.minim, optfcts.neldermead, optfcts.lmdif, optfcts.montecarlo]:
  ...     res = method(to_optimize, start, lo, hi)
  ...     if res[0]:
  ...         tbl.add_row([method.__name__, res[2], res[1][0], res[1][1]])
  ...     else:
  ...         print(f"Failed {method.__name__}: {res[3]}")
  ...
  Failed lmdif: improper input parameters
  >>> print(tbl)
      method           stat0                 x                  y
    ---------- --------------------- ------------------ ------------------
         minim  1.73832000295664e-11 1.9999999630648686  4.000644742396361
    neldermead 4.932797434506576e-11 2.0000007686008545 3.9994204823962214
    montecarlo 6.629393582604866e-12 1.9999997816942297  3.999564165982678

In this particular case :py:func:`~sherpa.optmethods.optfcts.lmdif`
failed, and this is because the ``to_optimize`` function returned None
as its second argument. For the other cases we can see that they all
found the minimum location.

In order to use :py:func:`~sherpa.optmethods.optfcts.lmdif` we need
the per-parameter statistic, which we can get with a small tweak::

  def to_optimize2(args):
      x = args[0]
      y = args[1]

>       pa = (a - x) * (a - b)
>       pb = b * (y - x * x) * (y - x * x)

Reviewer: ```suggestion
      pa = (a - x)**2
      pb = b * (y - x * x)**2
```
Again, just my personal preference.

      stat = pa * pa + pb * pb
      return stat, [pa, pb]

This lets us use `lmdif`::

  >>> res2 = optfcts.lmdif(to_optimize2, start, lo, hi)
  >>> res2[0]
  True
  >>> res2[1]
  [2.        4.0000846]

The `lmdif` optimizer is one of those that returns a number of
elements in the "extra" dictionary::

  >>> res2[4]
  {'info': 0, 'nfev': 108, 'covar': array([[ 1.56250000e-02, -1.04361202e-01],
         [-1.04361202e-01,  4.15098457e+03]]), 'num_parallel_map': 0}

Optimizing a model
------------------

We can re-create the :ref:`Fitting the NormGauss1D and Lorentz1D models <fit_peaked_data_normgauss1d_lorentz1d>`
section - at least for the :py:class:`~sherpa.models.basic.NormGauss1D` fit.

The normalized gaussian model can be expressed as

::

  def ngauss(x, ampl, pos, fwhm):
      term = 4 * np.log(2)

>       numerator = ampl * np.exp(-term * (x - pos) * (x - pos) / (fwhm * fwhm))

Reviewer: Again `**2` to reduce the line length? This is long in my browser window, but of course it depends on the size of your screen and settings.

      denominator = np.sqrt(np.pi / term) * fwhm
      return numerator / denominator

and the data we are going to fit is read from a file:

  >>> d = np.loadtxt('test_peak.dat')
  >>> x = d[:, 0]
  >>> y = d[:, 1]

> In this case the "callback" routine we want to give the minimizer a
> routine which will apply a least-squares minimizer to compare the

Reviewer: ```suggestion
In this case the "callback" routine we want to give the minimizer is a
routine which will apply a least-squares minimizer to compare the
```

Author: I'm re-wording this entire single-sentence paragraph.

model to the data (the `x` and `y` values for the independent and
dependent axes respectively).

::

  def cb(pars):
      model = ngauss(x, pars[0], pars[1], pars[2])
      delta = model - y
      statval = (delta * delta).sum()
      # Keep a record of the parameters we've visited
      store.add_row([statval] + list(pars))
      return statval, delta

.. note::

>    This function would normally be written as a closure to ensure

Reviewer: I don't know what a "closure" is, I would use a callable class that keep that info as a class attribute. But either way, the point is to point out that that's a simply example, so whatever works here.
Or maybe a callable class is just one example to implement a  "closure"? I need to google and learn that term!

Author: A closure "encloses" the symbols it uses and I have seen it used in python discussions - even Guido has used it ...

https://www.python.org/dev/peps/pep-3104/#id47

Author: I've rewritten this to say something "a class or a closure"

   that the use of global terms like `x`, `y`, `ngauss`, and `store` do not
   cause problems.

> For the starting point and bounds I'm using to use a "guess" based on

Reviewer: ```suggestion
For the starting point and bounds I'm using a "guess" based on
```

Reviewer: The rest of the test is written with "we do this". Now, you switch to "I". Intentional?

Author: I'm ambidextrous... We shoulf=d really scrub the documnetation for this sort of change in style - I do it a lot - but then I decide that consistency is a hobgoblin of minds that aren't knee deep in documentation and forget...
